File: GoTracer/src/Types/Sphere.go
package types

import "math"

type Sphere struct {
	Position Vector
	Radius   float64
	Surface  Surface
}

func (s Sphere) Intersects(ray Ray) *Intersection {
	eo := s.Position.Subtract(ray.Position)
	v := eo.Dot(ray.Direction)
	if v < 0.0 {
		return nil
	}
	dist := math.Pow(s.Radius, 2.0) - (eo.Dot(eo) - math.Pow(v, 2.0))
	if dist < 0.0 {
		return nil
	}
	dist = v - math.Sqrt(dist)
	intersection := Intersection{Ray: ray, Distance: dist, Object: s}
	return &intersection
}

func (s Sphere) Normal(position Vector) Vector {
	return position.Subtract(s.Position).Normalize()
}

func (s Sphere) GetSurface() Surface {
	return s.Surface
}

File: RustTracer/Raytracer/sphere.rs
use raytracer::scene_object::SceneObject;
use raytracer::vector::Vector;
use raytracer::ray::Ray;
use raytracer::intersection::Intersection;
use raytracer::surface::Surface;
use std::num::pow;

pub struct Sphere<'a> {
    pub position: Vector,
    pub radius : f64,
    pub surface : &'a Surface + 'a
}

impl<'a> SceneObject<'a> for Sphere<'a> {

	fn intersects(&'a self, ray : Ray) -> Option<(Intersection, &SceneObject<'a>)> {
		let eo = self.position.subtract(ray.position);
		let v = eo.dot(ray.direction);
		if v < 0.0 {
			None
		} else {
			let dist = pow(self.radius,2) - (eo.dot(eo) - pow(v, 2));
			if dist < 0.0 {
				None
			} else {
				let dist = v - dist.sqrt();
				let intersection = Intersection { ray: ray, distance:dist};
				Some((intersection, self as &SceneObject<'a>))
			}
		}
	}

	fn normal(&self, position : Vector) -> Vector {
		position.subtract(self.position).normalize()
	}

	fn get_surface(&'a self) -> &'a Surface +'a  {
		self.surface
	}
}